core/src/commonMain/kotlin/com/powersync/db/PowerSyncDatabaseImpl.kt

```diff
@@ -309,6 +309,62 @@ internal class PowerSyncDatabaseImpl(
         }
     }
 
+    override suspend fun getNextCrudTransactionBatch(transactionLimit: Int): CrudBatch? {
+        waitReady()
+        return internalDb.readTransaction { transaction ->
+            // Since tx_id can be null, we can't use a WHERE tx_id < ? with transactionLimit + first crud entry tx_id
+            // So we get all operations and group them by transaction or fall back to an individual transaction if tx_id is null
+            val allOperations =
```

Comment:
We would likely need an optimized query here that does the counting in SQL - fetching all items first could be too inefficient.

I know this isn't easy to get right in SQL. Some windowing magic might help, if you need suggestions we're also happy to help.